sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -420,6 +420,10 @@ object DateTimeUtils {
     Instant.ofEpochSecond(secs, mos * NANOS_PER_MICROS)
   }
 
+  def daysToInstant(daysSinceEpoch: SQLDate): Instant = {
+    Instant.ofEpochSecond(daysSinceEpoch * SECONDS_PER_DAY)
```

Comment:
You assume `daysSinceEpoch ` in UTC here which is incorrect, in general case.

Reply:
Can you elaborate? This function is the seconds equivalent of the one directly above:
```scala
  def microsToInstant(us: Long): Instant = {
    val secs = Math.floorDiv(us, MICROS_PER_SECOND)
    // Unfolded Math.floorMod(us, MICROS_PER_SECOND) to reuse the result of
    // the above calculation of `secs` via `floorDiv`.
    val mos = us - secs * MICROS_PER_SECOND
    Instant.ofEpochSecond(secs, mos * NANOS_PER_MICROS)
  }
```
I assume that the data is written as UTC, and we apply the timezone when reading:
```
  def convertTz(ts: SQLTimestamp, fromZone: ZoneId, toZone: ZoneId): SQLTimestamp = {
    val rebasedDateTime = microsToInstant(ts).atZone(toZone).toLocalDateTime.atZone(fromZone)
    instantToMicros(rebasedDateTime.toInstant)
  }
```
So if it is 1 hour after midnight, and you're in UTC-2, it will subtract a single day.

Reply:
`daysSinceEpoch` is "local" days. For example: 1 day is 1970-01-02, and in
- UTC is `Instant.ofEpochSecond(1 * SECONDS_PER_DAY)`
- `Europe/Amsterdam`
```scala
    assert(LocalDateTime.of(
      LocalDate.of(1970, 1, 2),
      LocalTime.MIDNIGHT)
      .atZone(ZoneId.of("Europe/Amsterdam"))
      .toInstant.toEpochMilli / MILLIS_PER_HOUR === SECONDS_PER_DAY/SECONDS_PER_HOUR)
23 did not equal 24
ScalaTestFailureLocation: org.apache.spark.sql.catalyst.util.DateTimeUtilsSuite at (DateTimeUtilsSuite.scala:664)
Expected :24
Actual   :23
```
The difference in 1 hour due to days are local.

Reply:
Catalyst's DATE type stores the number of days since 1970-01-01 independently from any time zone (NOT in UTC).

Reply:
So, if I understand it correctly, you don't want to compensate for the timezone while casting from a DoubleType to a DateType?

Reply:
I do believe it is not needed. To calculate average number of days, you can just do `Math.floorDiv(sum, count)`.

Reply:
Thanks for your input, less is more :)